### PR TITLE
Added win_find module

### PR DESCRIPTION
**Pull request by jborean93**
##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
win_find

##### ANSIBLE VERSION
```
ansible 2.2.0.0
  config file =
  configured module search path = Default w/o overrides
```

##### SUMMARY
Added win_find module that tries to replicate the existing find module but for Windows. This is a pull request after the repo merge for https://github.com/ansible/ansible-modules-extras/pull/3333

### CODE REVIEW

**Comment by mattclay** [comment]
@jborean93 The [tests](https://app.shippable.com/runs/584cda30bdca630f0022e978/1/console) are failing due to failing yamllint tests:

```
test/integration/targets/win_find/tasks/main.yml
  343:7     error    too many spaces before colon  (colons)
```

**Comment by jborean93** [comment]
Thanks @mattclay for the help, I've got a few other PRs I'm trying to get through and seem to be dealing with some weird issues where the hosts never become reachable but this should help me fix this one at least.

**Comment by mattclay** [comment]
@jborean93 We do have issues sometimes with EC2 instance startup taking too long. If that occurs, any Ansible member can restart your build for you. Just ping one of us on IRC (#ansible-devel on freenode), particularly myself or @gundalow.

**Comment by jborean93** [comment]
ready_for_review

**Comment by dagwieers** [comment]
I would like to see this merged before v2.3. Please rebase (and add them alphabetically) and then I will endorse it.

**Comment by jborean93** [comment]
@dagwieers has been rebased with the latest devel branch.

**Comment by dagwieers** [comment]
@jborean93 Alright, there are some minor things (parameter types, using $result dict instead of PSObject), but I prefer to get this merged first as it is already 2 months waiting. And then we can still do a PR to fix this once we know better what other syntactic changes we like in our modules. So I'd say:

shipit

**Review by dagwieers — APPROVED** [review]

**Comment by dagwieers** [comment]
But first fix the test ;-)
```json
An exception occurred during task execution. The full traceback is:
At C:\Users\Administrator\AppData\Local\Temp\ansible-tmp-1486505095.14-132697010756336\win_file.ps1:323 char:13

             Remove-Item -Recurse -Force $fileinfo

             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 fatal: [windows-2012-RTM]: FAILED! => {
     "changed": false, 
     "error_record": {
         "CategoryInfo": {
             "Activity": "Remove-Item", 
             "Category": 23, 
             "Reason": "DirectoryNotFoundException", 
             "TargetName": "C:\\ansible_testing", 
             "TargetType": "String"
         }, 
```
With error:
```json
"msg": "Could not find a part of the path 'C:\\ansible_testing\\win_find\\broken-link'."
```

**Comment by jborean93** [comment]
This one if very weird as the tests work on my local machine but for some reason the files on shippable are smaller and have a different checksum. Will need to look into this further.

**Comment by mattclay** [comment]
@jborean93 How much smaller? That sounds like it could be related to the changes made for PR #20876.

**Comment by jborean93** [comment]
@mattclay I think you are right I am still using an older version of ansible with the tests and doesn't have those changes introduced. Will see if I can test with the current version of ansible and rerun it locally.

**Comment by jborean93** [comment]
Thanks for the heads up @mattclay I've fixed up the tests. @dagwieers all ready for you to review when ready.

**Comment by jhawkesworth** [comment]
shipit

File: lib/ansible/modules/windows/win_find.ps1
Comment (dagwieers) [inline lib/ansible/modules/windows/win_find.ps1:36]: So something I would do (but it is not mandatory) is to add `-type` to everything.
- age_stamp could be `-type "int"`
- file_type could be `-type "str"`

We don't handle a list of paths yet, but it could be useful to run `Expand_Environment` on every path you consider to use, so people can take advantage of `%userprofile%` or `%public%`.

I would drop `-failifempty $false` because that's the default.

I would drop `-default $null` because that's the default.

I personally feel that only adding the non-default options is easier to see which options are e.g. required or have a default.

All of this we can do after this is merged, as I don't want delay merging any longer.

**Comment by dagwieers** [comment]
shipit

**Comment by jhawkesworth** [comment]
bot_status

**Comment by gundalow** [comment]
I believe the bot is having a little nap at the moment.

**Comment by gundalow** [comment]
@jborean93 What do you think of Dan's comments? Something to address now,  or look at in the future?

**Comment by jhawkesworth** [comment]
shipit

**Comment by jborean93** [comment]
@gundalow I am happy to raise another pull request but I am hoping to at least get something in first before the cut off time. I am also weary about committing more stuff and breaking the bot like 19145 seems to be.

**Comment by gundalow** [comment]
Merged, so you can now raise a PR for any follow up issues. Should the maintainer just be yourself? 

I'll update CHANGELOG.md tomorrow.

**Comment by jborean93** [comment]
@gundalow I'm find to be the maintainer, this should have also updated the CHANGELOG with the new module, not sure if anything else needs to change. Thanks for everyone's help with this much appreciated.

**Comment by jhawkesworth** [comment]
Yay!

**Comment by gundalow** [comment]
Maintainer set https://github.com/ansible/ansibullbot/pull/344
Changelog updated as part of this PR

Thanks everyone

**Pull request merged**
